libvast/src/table_slice.cpp

```diff
@@ -399,101 +393,204 @@ std::span<const std::byte> as_bytes(const table_slice& slice) noexcept {
+      co_return;
+    selection = evaluate(*tailored_expr, slice, selection);
+    // Do no rows qualify?
+    if (!any(selection))
+      co_return;
+  }
   // Do all rows qualify?
-  if (rank(xs_ids) == intersection_rank) {
-    result.emplace_back(slice);
-    return;
+  if (rank(selection) == slice.rows()) {
+    co_yield slice;
+    co_return;
   }
-  // Get the desired encoding, and the already serialized layout.
-  auto f = detail::overload{
-    []() noexcept -> table_slice_encoding {
-      die("cannot select from an invalid table slice");
-    },
-    [&](const auto& encoded) noexcept {
-      return builder_id(state(encoded, slice.state_)->encoding);
-    },
-  };
-  table_slice_encoding implementation_id
-    = visit(f, as_flatbuffer(slice.chunk_));
   // Start slicing and dicing.
-  auto builder
-    = factory<table_slice_builder>::make(implementation_id, slice.layout());
-  if (builder == nullptr) {
-    VAST_ERROR("{} failed to get a table slice builder for {}", __func__,
-               implementation_id);
-    return;
-  }
-  id last_offset = offset;
-  auto push_slice = [&] {
-    if (builder->rows() == 0)
-      return;
-    auto new_slice = builder->finish();
-    if (new_slice.encoding() == table_slice_encoding::none) {
-      VAST_WARN("{} got an empty slice", __func__);
-      return;
-    }
-    new_slice.offset(last_offset);
-    new_slice.import_time(slice.import_time());
-    result.emplace_back(std::move(new_slice));
-  };
-  auto flat_layout = flatten(caf::get<record_type>(slice.layout()));
-  auto last_id = last_offset - 1;
-  for (auto id : select(intersection)) {
-    // Finish last slice when hitting non-consecutive IDs.
-    if (last_id + 1 != id) {
-      push_slice();
-      last_offset = id;
-      last_id = id;
-    } else {
-      ++last_id;
-    }
-    VAST_ASSERT(id >= offset);
-    auto row = id - offset;
-    VAST_ASSERT(row < slice.rows());
-    for (size_t column = 0; column < flat_layout.num_fields(); ++column) {
-      auto cell_value = slice.at(row, column, flat_layout.field(column).type);
-      if (!builder->add(cell_value)) {
-        VAST_ERROR("{} failed to add data at column {} in row {} to the "
-                   "builder: {}",
-                   __func__, column, row, cell_value);
-        return;
-      }
-    }
+  auto batch = to_record_batch(slice);
+  for (const auto [first, last] : select_runs(selection)) {
+    auto selected = table_slice{
+      batch->Slice(detail::narrow_cast<int64_t>(first - offset),
+                   detail::narrow_cast<int64_t>(last - first)),
+      slice.layout(),
+    };
+    selected.offset(offset + first);
+    selected.import_time(slice.import_time());
+    co_yield std::move(selected);
   }
-  push_slice();
-}
-
-std::vector<table_slice>
-select(const table_slice& slice, const ids& selection) {
-  std::vector<table_slice> result;
-  select(result, slice, selection);
-  return result;
 }
 
 table_slice truncate(table_slice slice, size_t num_rows) {
```

Comment:
NB: Calling this function `head` (and the dual `tail`) would make for more intuitive semantics.

Reply:
Hah, I thought about renaming it to `head` yesterday when I touched it. I'll do that!

Reply:
Renamed it to `head`, and also introduced `tail` and then implemented `split` in terms of `head` and `tail`.